File: Common/TextWriterTimeStampAdder.vb
Namespace Common

    Public Class TextWriterTimeStampAdder
        Inherits IO.TextWriter

        Public Sub New(ByVal TextWriter As IO.TextWriter)
            Me.Base = TextWriter
        End Sub

        Private Function GetTimeStamp() As String
            Return Utilities.Representation.GetTimeStamp(False) & " :: "
        End Function

#Region "Writes Group"
        Public Overrides Sub Write(value As String)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(buffer() As Char, index As Integer, count As Integer)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Char)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(buffer() As Char)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(format As String, arg0 As Object)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(format As String, arg0 As Object, arg1 As Object)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(format As String, arg0 As Object, arg1 As Object, arg2 As Object)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(format As String, ParamArray arg() As Object)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Boolean)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Decimal)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Double)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Integer)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Long)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Object)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As Single)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As UInteger)
            Throw New NotSupportedException()
        End Sub

        Public Overrides Sub Write(value As ULong)
            Throw New NotSupportedException()
        End Sub
#End Region

#Region "WriteLines Group"
        Public Overrides Sub WriteLine(buffer() As Char, index As Integer, count As Integer)
            If buffer Is Nothing Then
                Me.Base.WriteLine(Me.GetTimeStamp())
                If Me.AutoFlush Then
                    Me.Base.Flush()
                End If
                Exit Sub
            End If

            Dim Stamp = Me.GetTimeStamp()
            Dim StampLength = Stamp.Length
            Me.Base.Write(Stamp)

            Dim SI = index
            Dim Bl = False
            For I As Integer = index To index + count
                Dim Ch As Char
                If I < index + count Then
                    Ch = buffer(I)
                Else
                    Ch = ControlChars.Lf
                End If
                If Ch = ControlChars.Cr Or Ch = ControlChars.Lf Then
                    If Bl Then
                        Me.Base.Write(New String(" "c, StampLength))
                    End If
                    Bl = True
                    Me.Base.WriteLine(buffer, SI, I - SI)

                    If I < index + count - 1 AndAlso
                       (Ch = ControlChars.Cr And buffer(I + 1) = ControlChars.Lf) Then
                        I += 1
                    End If
                    SI = I + 1
                End If
            Next

            If Me.AutoFlush Then
                Me.Base.Flush()
            End If
        End Sub

        Public Overrides Sub WriteLine(value As Object)
            Dim Formattable = TryCast(value, IFormattable)
            If Formattable IsNot Nothing Then
                Me.WriteLine(Formattable.ToString(Nothing, Me.Base.FormatProvider))
            Else
                Me.WriteLine(value?.ToString())
            End If
        End Sub

        Public Overrides Sub WriteLine()
            Me.WriteLine(Nothing, 0, 0)
        End Sub

        Public Overrides Sub WriteLine(buffer() As Char)
            Me.WriteLine(buffer, 0, buffer.Length)
        End Sub

        Public Overrides Sub WriteLine(value As String)
            Me.WriteLine(value?.ToCharArray(), 0, If(value?.Length, 0))
        End Sub

        Public Overrides Sub WriteLine(format As String, arg0 As Object)
            Me.WriteLine(String.Format(format, arg0))
        End Sub

        Public Overrides Sub WriteLine(format As String, arg0 As Object, arg1 As Object)
            Me.WriteLine(String.Format(format, arg0, arg1))
        End Sub

        Public Overrides Sub WriteLine(format As String, arg0 As Object, arg1 As Object, arg2 As Object)
            Me.WriteLine(String.Format(format, arg0, arg1, arg2))
        End Sub

        Public Overrides Sub WriteLine(format As String, ParamArray arg() As Object)
            Me.WriteLine(String.Format(format, arg))
        End Sub

        Public Overrides Sub WriteLine(value As Boolean)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As Char)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As Decimal)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As Double)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As Integer)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As Long)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As Single)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As UInteger)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub

        Public Overrides Sub WriteLine(value As ULong)
            Me.WriteLine(value.ToString(Me.Base.FormatProvider))
        End Sub
#End Region

#Region "Asyncs Group"
        Public Overrides Function WriteAsync(buffer() As Char, index As Integer, count As Integer) As Task
            Throw New NotSupportedException()
        End Function

        Public Overrides Function WriteAsync(value As Char) As Task
            Throw New NotSupportedException()
        End Function

        Public Overrides Function WriteAsync(value As String) As Task
            Throw New NotSupportedException()
        End Function

        Public Overrides Async Function WriteLineAsync(buffer() As Char, index As Integer, count As Integer) As Task
            If buffer Is Nothing Then
                Await Me.Base.WriteLineAsync(Me.GetTimeStamp())
                If Me.AutoFlush Then
                    Await Me.Base.FlushAsync()
                End If
                Exit Function
            End If

            Dim Stamp = Me.GetTimeStamp()
            Dim StampLength = Stamp.Length
            Await Me.Base.WriteAsync(Stamp)

            Dim SI = index
            Dim Bl = False
            For I As Integer = index To index + count
                Dim Ch As Char
                If I < index + count Then
                    Ch = buffer(I)
                Else
                    Ch = ControlChars.Lf
                End If
                If Ch = ControlChars.Cr Or Ch = ControlChars.Lf Then
                    If Bl Then
                        Await Me.Base.WriteAsync(New String(" "c, StampLength))
                    End If
                    Bl = True
                    Await Me.Base.WriteLineAsync(buffer, SI, I - SI)

                    If I < index + count - 1 AndAlso
                       (Ch = ControlChars.Cr And buffer(I + 1) = ControlChars.Lf) Then
                        I += 1
                    End If
                    SI = I + 1
                End If
            Next

            If Me.AutoFlush Then
                Await Me.Base.FlushAsync()
            End If
        End Function

        Public Overrides Async Function WriteLineAsync() As Task
            Await Me.WriteLineAsync(Nothing, 0, 0)
        End Function

        Public Overrides Async Function WriteLineAsync(value As Char) As Task
            Await Me.WriteLineAsync(value.ToString(Me.Base.FormatProvider))
        End Function

        Public Overrides Async Function WriteLineAsync(value As String) As Task
            Await Me.WriteLineAsync(value?.ToCharArray(), 0, If(value?.Length, 0))
        End Function
#End Region

        Public Overrides Sub Close()
            Me.Base.Close()
        End Sub

        Public Overrides Function CreateObjRef(requestedType As Type) As Runtime.Remoting.ObjRef
            Return Me.Base.CreateObjRef(requestedType)
        End Function

        Protected Overrides Sub Dispose(disposing As Boolean)
            Me.Base.Dispose()
        End Sub

        Public Overrides Sub Flush()
            Me.Base.Flush()
        End Sub

        Public Overrides Function FlushAsync() As Task
            Return Me.Base.FlushAsync()
        End Function

        Public Overrides Function InitializeLifetimeService() As Object
            Return Me.Base.InitializeLifetimeService()
        End Function

        Public Overrides Function ToString() As String
            Return String.Concat(NameOf(TextWriterTimeStampAdder), "{", Me.Base.ToString(), "}")
        End Function

        Public Overrides ReadOnly Property Encoding As Text.Encoding
            Get
                Return Me.Base.Encoding
            End Get
        End Property

        Public Overrides Property NewLine As String
            Get
                Return Me.Base.NewLine
            End Get
            Set(value As String)
                Me.Base.NewLine = value
            End Set
        End Property

        Public Overrides ReadOnly Property FormatProvider As IFormatProvider
            Get
                Return Me.Base.FormatProvider
            End Get
        End Property

#Region "AutoFlush Property"
        Private _AutoFlush As Boolean = True

        Public Property AutoFlush As Boolean
            Get
                Return Me._AutoFlush
            End Get
            Set(ByVal Value As Boolean)
                Me._AutoFlush = Value
            End Set
        End Property
#End Region

        Private ReadOnly Base As IO.TextWriter

    End Class

End Namespace

File: Common/TextWriterTimeStampAdder.cs
using System.Threading.Tasks;
using Microsoft.VisualBasic;
using System;

namespace Ks
{
    namespace Common
    {
        public class TextWriterTimeStampAdder : System.IO.TextWriter
        {
            public TextWriterTimeStampAdder(System.IO.TextWriter TextWriter, Func<string> GetTimeStampDelegate)
            {
                this.Base = TextWriter;
                this.GetTimeStampDelegate = GetTimeStampDelegate;
            }

            public TextWriterTimeStampAdder(System.IO.TextWriter TextWriter) : this(TextWriter, () => Utilities.Representation.GetTimeStamp())
            {
            }

            private string GetTimeStamp()
            {
                return this.GetTimeStampDelegate.Invoke() + " :: ";
            }

            public override void Write(string value)
            {
                throw new NotSupportedException();
            }

            public override void Write(char[] buffer, int index, int count)
            {
                throw new NotSupportedException();
            }

            public override void Write(char value)
            {
                throw new NotSupportedException();
            }

            public override void Write(char[] buffer)
            {
                throw new NotSupportedException();
            }

            public override void Write(string format, object arg0)
            {
                throw new NotSupportedException();
            }

            public override void Write(string format, object arg0, object arg1)
            {
                throw new NotSupportedException();
            }

            public override void Write(string format, object arg0, object arg1, object arg2)
            {
                throw new NotSupportedException();
            }

            public override void Write(string format, params object[] arg)
            {
                throw new NotSupportedException();
            }

            public override void Write(bool value)
            {
                throw new NotSupportedException();
            }

            public override void Write(decimal value)
            {
                throw new NotSupportedException();
            }

            public override void Write(double value)
            {
                throw new NotSupportedException();
            }

            public override void Write(int value)
            {
                throw new NotSupportedException();
            }

            public override void Write(long value)
            {
                throw new NotSupportedException();
            }

            public override void Write(object value)
            {
                throw new NotSupportedException();
            }

            public override void Write(float value)
            {
                throw new NotSupportedException();
            }

            public override void Write(uint value)
            {
                throw new NotSupportedException();
            }

            public override void Write(ulong value)
            {
                throw new NotSupportedException();
            }

            public override void WriteLine(char[] buffer, int index, int count)
            {
                if (buffer == null)
                {
                    this.Base.WriteLine(this.GetTimeStamp());
                    if (this.AutoFlush)
                        this.Base.Flush();
                    return;
                }

                var Stamp = this.GetTimeStamp();
                var StampLength = Stamp.Length;
                this.Base.Write(Stamp);

                var SI = index;
                var Bl = false;
                var loopTo = index + count;
                for (int I = index; I <= loopTo; I++)
                {
                    char Ch = default(char);
                    if (I < (index + count))
                        Ch = buffer[I];
                    else
                        Ch = ControlChars.Lf;
                    if ((Ch == ControlChars.Cr) | (Ch == ControlChars.Lf))
                    {
                        if (Bl)
                            this.Base.Write(new string(' ', StampLength));
                        Bl = true;
                        this.Base.WriteLine(buffer, SI, I - SI);

                        if (I < ((index + count) - 1) && ((Ch == ControlChars.Cr) & (buffer[I + 1] == ControlChars.Lf)))
                            I += 1;
                        SI = I + 1;
                    }
                }

                if (this.AutoFlush)
                    this.Base.Flush();
            }

            public override void WriteLine(object value)
            {
                var Formattable = value as IFormattable;
                if (Formattable != null)
                    this.WriteLine(Formattable.ToString(null, this.Base.FormatProvider));
                else
                    this.WriteLine(value?.ToString());
            }

            public override void WriteLine()
            {
                this.WriteLine(null, 0, 0);
            }

            public override void WriteLine(char[] buffer)
            {
                this.WriteLine(buffer, 0, buffer.Length);
            }

            public override void WriteLine(string value)
            {
                this.WriteLine(value?.ToCharArray(), 0, value?.Length ?? 0);
            }

            public override void WriteLine(string format, object arg0)
            {
                this.WriteLine(string.Format(format, arg0));
            }

            public override void WriteLine(string format, object arg0, object arg1)
            {
                this.WriteLine(string.Format(format, arg0, arg1));
            }

            public override void WriteLine(string format, object arg0, object arg1, object arg2)
            {
                this.WriteLine(string.Format(format, arg0, arg1, arg2));
            }

            public override void WriteLine(string format, params object[] arg)
            {
                this.WriteLine(string.Format(format, arg));
            }

            public override void WriteLine(bool value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(char value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(decimal value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(double value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(int value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(long value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(float value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(uint value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override void WriteLine(ulong value)
            {
                this.WriteLine(value.ToString(this.Base.FormatProvider));
            }

            public override Task WriteAsync(char[] buffer, int index, int count)
            {
                throw new NotSupportedException();
            }

            public override Task WriteAsync(char value)
            {
                throw new NotSupportedException();
            }

            public override Task WriteAsync(string value)
            {
                throw new NotSupportedException();
            }

            public override async Task WriteLineAsync(char[] buffer, int index, int count)
            {
                if (buffer == null)
                {
                    await this.Base.WriteLineAsync(this.GetTimeStamp());
                    if (this.AutoFlush)
                        await this.Base.FlushAsync();
                    return;
                }

                var Stamp = this.GetTimeStamp();
                var StampLength = Stamp.Length;
                await this.Base.WriteAsync(Stamp);

                var SI = index;
                var Bl = false;
                var loopTo = index + count;
                for (int I = index; I <= loopTo; I++)
                {
                    char Ch = default(char);
                    if (I < (index + count))
                        Ch = buffer[I];
                    else
                        Ch = ControlChars.Lf;
                    if ((Ch == ControlChars.Cr) | (Ch == ControlChars.Lf))
                    {
                        if (Bl)
                            await this.Base.WriteAsync(new string(' ', StampLength));
                        Bl = true;
                        await this.Base.WriteLineAsync(buffer, SI, I - SI);

                        if (I < ((index + count) - 1) && ((Ch == ControlChars.Cr) & (buffer[I + 1] == ControlChars.Lf)))
                            I += 1;
                        SI = I + 1;
                    }
                }

                if (this.AutoFlush)
                    await this.Base.FlushAsync();
            }

            public override async Task WriteLineAsync()
            {
                await this.WriteLineAsync(null, 0, 0);
            }

            public override async Task WriteLineAsync(char value)
            {
                await this.WriteLineAsync(value.ToString(this.Base.FormatProvider));
            }

            public override async Task WriteLineAsync(string value)
            {
                await this.WriteLineAsync(value?.ToCharArray(), 0, value?.Length ?? 0);
            }

            public override void Close()
            {
                this.Base.Close();
            }

            public override System.Runtime.Remoting.ObjRef CreateObjRef(Type requestedType)
            {
                return this.Base.CreateObjRef(requestedType);
            }

            protected override void Dispose(bool disposing)
            {
                this.Base.Dispose();
            }

            public override void Flush()
            {
                this.Base.Flush();
            }

            public override Task FlushAsync()
            {
                return this.Base.FlushAsync();
            }

            public override object InitializeLifetimeService()
            {
                return this.Base.InitializeLifetimeService();
            }

            public override string ToString()
            {
                return string.Concat(nameof(TextWriterTimeStampAdder), "{", this.Base.ToString(), "}");
            }

            public override System.Text.Encoding Encoding
            {
                get
                {
                    return this.Base.Encoding;
                }
            }

            public override string NewLine
            {
                get
                {
                    return this.Base.NewLine;
                }
                set
                {
                    this.Base.NewLine = value;
                }
            }

            public override IFormatProvider FormatProvider
            {
                get
                {
                    return this.Base.FormatProvider;
                }
            }

            private bool _AutoFlush = true;

            public bool AutoFlush
            {
                get
                {
                    return this._AutoFlush;
                }
                set
                {
                    this._AutoFlush = value;
                }
            }

            private readonly System.IO.TextWriter Base;
            private readonly Func<string> GetTimeStampDelegate;
        }
    }
}
